### DISCUSSION TITLE
Invoking a callback after popover closes

### DISCUSSION THREAD

**Discussion by marcospassos** [ariakit/ariakit#1638] [Questions] [answered]
We have a use case where clicking on a menu option changes the order of the parent element, causing a visual shift (the popup is repositioned while closing):

https://user-images.githubusercontent.com/943036/179579399-f9427a33-7ca3-4db8-a800-2fa963a50ded.mov

One solution is applying a delay, so the action is only executed after the menu closes. This is a very fragile solution, though, as the delay needs to match the duration of the animation, and any change to either side would break the experience.

So I'm wondering whether somehow I could "schedule" menu actions to execute after the menu closes or somehow keep the menu at the last position before closing.

Is there any API that I could take advantage of?

**Comment by diegohaz** [comment, marked as answer]
> So I'm wondering whether somehow I could "schedule" menu actions to execute after the menu closes or somehow keep the menu at the last position before closing.

The menu state has a `mounted` property that's only set to `false` when the animation finishes. You can set a local state in your MenuItem onClick handler and handle the action in an effect.

Something like this:

```jsx
const [action, setAction] = useState(null);

useEffect(() => {
  if (!action) return;
  if (menu.mounted) return;
  action();
  setAction(null);
}, [action, menu.mounted]);

<MenuItem onClick={() => setAction(() => () => reorder())} />
```

As for preventing the menu from being repositioned, I'm not sure this works, but you can probably achieve it with `getAnchorRect`.:

```jsx
useMenuState({
  getAnchorRect: (anchor) => {
    if (justReordered) {
      return previousAnchorRect;
    }
    return anchor.getBoundingClientRect();
  },
});
```

**Reply by marcospassos** [reply]
> ```js
> const [action, setAction] = useState(null);
> 
> useEffect(() => {
>   if (!action) return;
>   if (menu.mounted) return;
>   action();
>   setAction(null);
> }, [action, menu.mounted]);
> 
> <MenuItem onClick={() => setAction(() => () => reorder())} />
> ```

This worked just great! Thank you!
